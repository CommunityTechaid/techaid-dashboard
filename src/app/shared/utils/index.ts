export { LRUMap } from './lru';
export { sha1_hash } from './sha1';
export { Tree, Node } from './tree_model';
export { HashUtils, Token } from './hash_utils';
export { DateUtils } from './date_utils';