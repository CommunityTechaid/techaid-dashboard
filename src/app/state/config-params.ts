export interface ConfigParams {
    production: boolean
    graphql_endpoint: string
    environment: string
    auth_enabled: boolean
    remote_config: boolean
    auth_endpoint?: string
    version: any
    pusher?: any
}