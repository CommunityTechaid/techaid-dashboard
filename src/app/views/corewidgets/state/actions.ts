export class SearchQuery {
    static readonly type = '[Auth Server] Search Query';
    constructor(public query: string) { }
}
