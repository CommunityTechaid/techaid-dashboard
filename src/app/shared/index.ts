import { NgModule, ModuleWithProviders } from '@angular/core';
import { Routes, RouterModule } from "@angular/router";
import { BootstrapStatusDirective } from './directives/bootstrap-status.directive';
import { InputMaskComponent } from './components/input-mask/input-mask.component';
import { ImagePreloadDirective } from './directives/image-preload.directive';
import { AutoScroll } from './directives/autoscroll.directive';
import { AuthGuard } from './services/auth.guard';
import { ConfigService } from './services/config.service';
import { PusherService } from './services/pusher.service';
import { PingService } from './services/ping.service';
import { TimeAgoPipe } from './pipes/timeago.pipe';
import { CodeHighlightService } from './services/code-highlight.service';
import { AppInitialComponent } from './components/app-initial/app-initial.component';
import { AuthenticationService } from './services/authentication.service';

@NgModule({
    imports: [
    ],
    exports: [
        BootstrapStatusDirective,
        InputMaskComponent,
        TimeAgoPipe,
        AutoScroll,
        ImagePreloadDirective,
        AppInitialComponent
    ],
    declarations: [
        BootstrapStatusDirective,
        InputMaskComponent,
        ImagePreloadDirective,
        TimeAgoPipe,
        AppInitialComponent,
        AutoScroll
    ]
})
export class PicSharedModule {
    static forRoot(): ModuleWithProviders {
        return {
            ngModule: PicSharedModule,
            providers: [
                AuthenticationService,
                AuthGuard,
                ConfigService,
                PusherService,
                PingService,
                CodeHighlightService
            ]
        }
    }
}
