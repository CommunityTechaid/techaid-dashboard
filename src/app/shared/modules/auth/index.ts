import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { Routes, RouterModule } from '@angular/router';
import { ReactiveFormsModule } from '@angular/forms';
import { PicSharedModule } from '@app/shared'
import { HTTP_INTERCEPTORS } from '@angular/common/http';

export const ROUTES: Routes = [
]

@NgModule({
    imports: [
        CommonModule,
        ReactiveFormsModule,
        RouterModule.forChild(ROUTES),
        PicSharedModule
    ],
    declarations: [
    ],
    providers: [
        /*
        {
            provide: HTTP_INTERCEPTORS,
            useClass: TokenInterceptor,
            multi: true
        }
        */
    ]
})
export class AppAuthModule {}
