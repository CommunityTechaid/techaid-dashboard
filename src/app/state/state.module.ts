import { NgModule, ModuleWithProviders } from '@angular/core';
import { NgxsModule } from '@ngxs/store';
import { NgxsLoggerPluginModule } from '@ngxs/logger-plugin';
import { NgxsReduxDevtoolsPluginModule } from '@ngxs/devtools-plugin';
import { NgxsRouterPluginModule } from '@ngxs/router-plugin';
import { NgxsFormPluginModule } from '@ngxs/form-plugin';
import { UserState } from './user/user.state';
export { UserState } from './user/user.state';
const DevToolsOptions: any = {
    name: 'PICNGXS',
    serialize: undefined
}

@NgModule({
    exports: [
    ],
    imports: [
        NgxsModule.forRoot([UserState]),
        //NgxsLoggerPluginModule.forRoot(),
        NgxsRouterPluginModule.forRoot(),
        NgxsFormPluginModule.forRoot(),
        NgxsReduxDevtoolsPluginModule.forRoot()
    ],
})
export class PicStateModule {
    static forRoot() {
        return {
            ngModule: PicStateModule,
            providers: [
            ]
        }
    }
}
