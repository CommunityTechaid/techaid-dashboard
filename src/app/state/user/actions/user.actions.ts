import { User } from '../user.state';

export class LoginUser {
    static readonly type = '[User API] Login User';
    constructor() { }
}

export class LogoutUser {
    static readonly type = '[User API] Logout User';
    constructor() { }
}
