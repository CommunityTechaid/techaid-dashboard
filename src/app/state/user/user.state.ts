import { State, StateContext, Action, Selector, NgxsOnInit } from '@ngxs/store';
import { LoginUser, LogoutUser } from './actions/user.actions';
import { NgZone, Injectable } from '@angular/core';
import { AuthenticationService } from '@app/shared/services/authentication.service';
import { tap, first } from 'rxjs/operators';

export interface UserStateModel {
   user: User;
}

export class User {
    name: String;
    picture: String;
    authenticated: Boolean;
    permissions: Array<string> = [];
    authorities: {[key: string]: boolean} = {};

    constructor(init?: Partial<User>) {
        Object.assign(this, init);
        (this.permissions || []).forEach(p => {
            this.authorities[p] = true;
        });
    }
}

@State<UserStateModel>({
    name: 'user',
    defaults: {
        user: new User({
            name: '',
            picture: '',
            authenticated: false,
            permissions: [],
            authorities: {}
        })
    }
})
@Injectable()
export class UserState implements NgxsOnInit {
    constructor(private auth: AuthenticationService, private zone: NgZone) { }

    @Selector() static user(state: UserStateModel) {
        return state.user;
    }

    ngxsOnInit(ctx: StateContext<UserStateModel>) {
        this.auth.isAuthenticated$.pipe(first()).subscribe(loggedIn => {
            if (loggedIn) {
                this.handleLoggedIn(ctx);
            }
        });
    }

    handleLoggedIn(ctx: StateContext<UserStateModel>) {
        ctx.patchState({
            user: new User({
                name: 'Me',
                picture: '',
                authenticated: true,
            })
        });
        this.auth.getUser$().subscribe(u => {
            u.authenticated = true;
            ctx.patchState({user: new User(u)});
            this.auth.getTokenSilently$().pipe(first()).subscribe(data => {
                const parts = data.split('.');
                const token = JSON.parse(atob(parts[1]));
                u.permissions = token.permissions || [];
                ctx.patchState({user: new User(u)});
            });
        });
    }


    @Action(LoginUser)
    loginUser(ctx: StateContext<UserStateModel>, action: LoginUser) {
        this.auth.isAuthenticated$.pipe(first()).subscribe(loggedIn => {
            if (loggedIn) {
                this.handleLoggedIn(ctx);
            } else {
                this.auth.login();
            }
        });
    }

    @Action(LogoutUser)
    logoutUser(ctx: StateContext<UserStateModel>, action: LogoutUser) {
        const state = ctx.getState();
        ctx.patchState({
            user: new User({
                authenticated: false,
                name: '',
                picture: '',
            })
        });

        this.zone.run(() => {
            this.auth.logout();
        });
    }
}
