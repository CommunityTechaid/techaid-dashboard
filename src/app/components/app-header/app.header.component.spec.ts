import { TestBed, async, ComponentFixture } from '@angular/core/testing';
import { AppHeader } from './app.header.component';
import { AppSharedModule } from '@app/shared';
import { NO_ERRORS_SCHEMA } from '@angular/core';
import { NoopAnimationsModule } from '@angular/platform-browser/animations';
import { HttpClientModule } from '@angular/common/http';
import { AppStateModule } from '@app/state/state.module';
import { RouterTestingModule } from '@angular/router/testing';
import { RouterModule } from '@angular/router';
import { APP_BASE_HREF } from '@angular/common';
import { Store } from '@ngxs/store';

import { LoginUser, LogoutUser } from '@app/state/user/actions/user.actions';
import { User } from '@app/shared/services/auth.service';
import { AuthService } from '@app/shared/services/auth.service';
import { MockAuthService } from '@app/shared/mocks/mock.auth.service';

describe('AppHeader', () => {
  let fixture: ComponentFixture<AppHeader>;
  let component: AppHeader;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [
        AppHeader
      ],
      imports: [
        AppStateModule,
        RouterTestingModule,
        NoopAnimationsModule
      ],
      providers: [{ provide: AuthService, useClass: MockAuthService }],
      schemas: [NO_ERRORS_SCHEMA]
    }).compileComponents();

    fixture = TestBed.createComponent(AppHeader);
    component = fixture.componentInstance;
    fixture.detectChanges();
  }));

  it('renders user logged in state', async(() => {
    const store: Store = TestBed.get(Store);
    const user: User = new User({
      username: 'mock',
      name: 'Mocked User',
      token: 'logged_in'
    });

    const compiled = fixture.debugElement.nativeElement;
    expect(compiled.querySelector('.navbar-nav a.nav-item').textContent).toContain('Login');

    store.dispatch(new LoginUser(user)).subscribe(() => {
      fixture.detectChanges();
      expect(component.user.username).toEqual(user.username);
      expect(compiled.querySelector('.navbar-nav a.nav-item').textContent).toContain('Mocked User');

      store.dispatch(new LogoutUser()).subscribe(() => {
        fixture.detectChanges();
        expect(component.user).toEqual({});
        expect(compiled.querySelector('.navbar-nav a.nav-item').textContent).toContain('Login');
      });
    });
  }));


  it('renders markup to snapshot', async(() => {
    expect(fixture).toMatchSnapshot();
  }));
});
