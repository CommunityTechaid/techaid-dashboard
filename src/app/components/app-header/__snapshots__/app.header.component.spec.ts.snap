// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppHeader renders markup to snapshot 1`] = `
<app-header
  __user$__selector={[Function Observable]}
  store={[Function Store]}
  sub={[Function Subscriber]}
  user={[Function User]}
>
  <nav
    _ngcontent-c1=""
    class="navbar navbar-expand-lg navbar-light fixed-top bg-white nav-border"
  >
    <div
      _ngcontent-c1=""
      class="container-fluid"
    >
      <a
        _ngcontent-c1=""
        class="navbar-brand"
        href="/"
        ng-reflect-router-link="/"
      >
        <img
          _ngcontent-c1=""
          alt=""
          src="assets/logo.jpg"
          width="200px"
        />
      </a>
      <span
        _ngcontent-c1=""
      >
        <sup
          _ngcontent-c1=""
          class="text-muted"
          title="null"
        >
          v
        </sup>
      </span>
      <button
        _ngcontent-c1=""
        aria-controls="navbarNavAltMarkup"
        aria-expanded="false"
        aria-label="Toggle navigation"
        class="navbar-toggler"
        data-target="#navbarNavAltMarkup"
        data-toggle="collapse"
        type="button"
      >
        <span
          _ngcontent-c1=""
          class="navbar-toggler-icon"
        />
        _ 
      </button>
      <div
        _ngcontent-c1=""
        class="collapse navbar-collapse"
      >
        <ul
          _ngcontent-c1=""
          class="navbar-nav mr-auto"
        />
        <div
          _ngcontent-c1=""
          class="navbar-nav"
        >
          
          <a
            _ngcontent-c1=""
            class="nav-item nav-link"
            href="/login"
            style="margin-left: auto"
          >
            <i
              _ngcontent-c1=""
              class="fa fa-fw fa-sign-in"
            />
             Â  Login 
          </a>
          
        </div>
      </div>
    </div>
  </nav>
</app-header>
`;
